### PR TITLE
DT-2563: add endpoint to get role details from hmpps-auth

### PR DESCRIPTION
**Pull request by SimonMitchellMOJ**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/config/ManageUsersApiExceptionHandler.kt [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+    private val log = LoggerFactory.getLogger(this::class.java)
+  }
+}
+
+data class ErrorResponse(
+  val status: Int,
+  val errorCode: Int? = null,
+  val userMessage: String? = null,
+  val developerMessage: String? = null,
+  val moreInfo: String? = null
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/config/ManageUsersApiExceptionHandler.kt:109]: maybe ditch `moreInfo` doesn't look like we ever use it

File: build.gradle.kts [hunk 2]
```diff
@@ -8,7 +8,25 @@ configurations {
 }
 
 dependencies {
+  implementation("org.springframework.boot:spring-boot-starter-oauth2-resource-server")
+  implementation("org.springframework.boot:spring-boot-starter-security")
   implementation("org.springframework.boot:spring-boot-starter-webflux")
+  implementation("org.springframework.boot:spring-boot-starter-oauth2-client")
+  implementation("org.apache.commons:commons-lang3:3.12.0")
+
+  implementation("com.google.code.gson:gson:2.8.8")
```
Comment (petergphillips) [inline build.gradle.kts:17]: are you using gson?

File: build.gradle.kts [hunk 3]
```diff
@@ -8,7 +8,25 @@ configurations {
 }
 
 dependencies {
+  implementation("org.springframework.boot:spring-boot-starter-oauth2-resource-server")
+  implementation("org.springframework.boot:spring-boot-starter-security")
   implementation("org.springframework.boot:spring-boot-starter-webflux")
+  implementation("org.springframework.boot:spring-boot-starter-oauth2-client")
+  implementation("org.apache.commons:commons-lang3:3.12.0")
```
Comment (petergphillips) [inline build.gradle.kts:15]: is this needed?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/resource/RolesController.kt [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+@Validated
+@RestController
+class RolesController(
+  private val rolesService: RolesService
+) {
+
+  @PreAuthorize("hasRole('ROLE_ROLES_ADMIN')")
+  @Operation(
+    summary = "Get all audit events",
+    description = "Get all audit events, role required is ROLE_AUDIT",
```
Comment (petergphillips) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/resource/RolesController.kt:27]: description here doesn't match `PreAuthorize` requirement

File: src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/resource/RolesController.kt [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+@Validated
+@RestController
+class RolesController(
+  private val rolesService: RolesService
+) {
+
+  @PreAuthorize("hasRole('ROLE_ROLES_ADMIN')")
+  @Operation(
+    summary = "Get all audit events",
```
Comment (petergphillips) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/resource/RolesController.kt:26]: copied from an audit controller :thinking:

File: src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/utils/UserContextFilter.kt [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+import javax.servlet.Filter
+import javax.servlet.FilterChain
+import javax.servlet.FilterConfig
+import javax.servlet.ServletException
+import javax.servlet.ServletRequest
+import javax.servlet.ServletResponse
+import javax.servlet.http.HttpServletRequest
+
+@Component
+@Order(4)
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/utils/UserContextFilter.kt:16]: I wonder what this annotation gives us?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/resource/RolesController.kt [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+    description = "Get all audit events, role required is ROLE_AUDIT",
+    security = [SecurityRequirement(name = "ROLE_ROLES_ADMIN")],
+    responses = [
+      ApiResponse(
+        responseCode = "200",
+        description = "Role Details Returned",
+        content = [
+          Content(
+            mediaType = "application/json",
+            array = ArraySchema(schema = Schema(implementation = Role::class))
```
Comment (petergphillips) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/resource/RolesController.kt:36]: do you actually need this line, would have thought it should be calculated from the method signature?

File: src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/resource/RolesControllerIntTest.kt [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+    webTestClient.get().uri("/roles/role-code")
+      .exchange()
+      .expectStatus().isUnauthorized
+  }
+
+  @Test
+  fun `access forbidden when no role`() {
+
+    webTestClient.get().uri("/roles/AUTH_GROUP_MANAGER")
```
Comment (andymarke) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/resource/RolesControllerIntTest.kt:20]: maybe an additional test where you supply the wrong role?

**Review by andymarke — APPROVED** [review]
LGTM

File: src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/config/ResourceServerConfiguration.kt [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+class ResourceServerConfiguration : WebSecurityConfigurerAdapter() {
+  override fun configure(http: HttpSecurity) {
+    http
+      .sessionManagement()
+      .sessionCreationPolicy(SessionCreationPolicy.STATELESS)
+      .and().csrf().disable()
+      .authorizeRequests { auth ->
+        auth.antMatchers(
+          "/webjars/**", "/favicon.ico", "/csrf",
+          "/health/**", "/info", "/ping",
```
Comment (petergphillips) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/config/ResourceServerConfiguration.kt:22]: don't need `/ping` here as is at `/health/ping` these days

File: src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/config/OpenApiConfiguration.kt [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+
+@Configuration
+class OpenApiConfiguration(buildProperties: BuildProperties) {
+  private val version: String = buildProperties.version
+
+  @Bean
+  fun customOpenAPI(): OpenAPI = OpenAPI()
+    .servers(
+      listOf(
+        Server().url("https://hmpps-manage-user-api.hmpps.service.justice.gov.uk").description("Prod"),
```
Comment (petergphillips) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/config/OpenApiConfiguration.kt:19]: think you're missing a `s` in `manage-users-api`?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/utils/UserContextFilter.kt (hunk 6 above)
Comment (SimonMitchellMOJ) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/utils/UserContextFilter.kt:16]: If I remember correctly Spring Security doesn't set an order on the Filter bean that it creates so by adding the order it just means they start the same each time, In some of the other projects there are more filters and I think one relied on an element in another so needed to be specifically ordered - will update this filter to number Order(2)

**Pull request merged**